ratelimit_test.go

```diff
@@ -189,22 +189,91 @@ func TestPer(t *testing.T) {
 }
 
 func TestSlack(t *testing.T) {
-	runTest(t, func(r testRunner) {
-		slow := r.createLimiter(10, WithoutSlack)
-		// Defaults to 10 slack.
-		fast := r.createLimiter(100)
+	// To simulate slack, we combine two limiters.
+	// - First, we start a single goroutine with both of them,
+	//   during this time the slow limiter will dominate,
+	//   and allow the fast limiter to acumulate slack.
+	// - After 2 seconds, we start another goroutine with
+	//   only the faster limiter. This will allow it to max out,
+	//   and consume all the slack.
+	// - After 3 seconds, we look at the final result, and we expect,
+	//   a sum of:
+	//   - slower limiter running for 3 seconds
+	//   - faster limiter running for 1 second
+	//   - slack accumulated by the faster limiter during the two seconds.
+	//     it was blocked by slower limiter.
+	tests := []struct {
```

Comment:
I also had the same test generated dynamically:

```lang=go
	const (
		slowRate = 10
		fastRate = 100
	)

	// Test for a few Per settings.
	testsPer := []time.Duration{
		// Default.
		{time.Second},
		{time.Millisecond * 500},
	}
	// And a few different slack options
	testsSlack := []int{10, 20, 100, 150}

	for _, per := range testsPer {
		for _, slack := range testsSlack {
			testKey := fmt.String("per-%v-slack-%v", per, slack)
			t.Run(testKey, func(t *testing.T) {
				runTest(t, func(r testRunner) {
					slow := r.createLimiter(slowRate, WithoutSlack)
					fast := r.createLimiter(fastRate, tt.opt...)

					r.startTaking(slow, fast)

					r.afterFunc(2, func() {
						r.startTaking(fast)
						r.startTaking(fast)
					})

					// slow limiter with 10hz dominates here - we're always at it's limit.
					r.assertCountAt(1*time.Second, slowRate)

					// results:
					// - for two seconds we ran a slow rate
					// - for a second we ran a fast rate
					// - and we add the slack
					// rates have to be adjusted for Per
					adjustedSlow := slowRate * int(time.Second/testsPer)
					asjustedFast := fastRate * int(time.Second/testsPer)
					r.assertCountAt(3*time.Second, asjustedSlow+adjustedFast+slack)
				})
			})
		}
	}
```

but it seems like manual definition is easier to understand at this point.